### PR TITLE
JAMES-1692 setMessages, Sending a message : save a new message to outbox

### PR DESCRIPTION
**Pull request by fvn-linagora**
As a jmap client, when sending new messages via setMessages, I should get a response containing saved messages, including set IDs.

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MessageWithId.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.methods;
+
+import org.apache.james.jmap.model.CreationMessage;
+
+class MessageWithId<T> {
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MessageWithId.java]: visibility?

File: server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java [hunk 2]
```diff
@@ -54,6 +61,16 @@ protected void configure() {
         methods.addBinding().to(new TypeLiteral<GetMessagesMethod<CassandraId>>(){});
         methods.addBinding().to(new TypeLiteral<SetMessagesMethod<CassandraId>>(){});
         bind(SetMessagesUpdateProcessor.class).to(new TypeLiteral<SetMessagesUpdateProcessor<CassandraId>>(){});
```
Comment (aduprat) [inline server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java]: SetMessagesDestructionProcessor missing.

Comment (aduprat) [inline server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java]: You should define the scope of the SetMessages*Processors.

File: server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java [hunk 3]
```diff
@@ -54,6 +61,16 @@ protected void configure() {
...
+        bind(SetMessagesCreationProcessor.class).to(new TypeLiteral<SetMessagesCreationProcessor<CassandraId>>(){});
+    }
+
+    @Provides
```
Comment (aduprat) [inline server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java]: Not needed, you may use a Multibinder on SetMessagesProcessor instead.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/MailboxRoleNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.exceptions;
+
+import org.apache.james.jmap.model.mailbox.Role;
+
+public class MailboxRoleNotFoundException extends RuntimeException {
+
+    final Role role;
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/MailboxRoleNotFoundException.java]: private

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            outbox = getOutbox(mailboxSession).orElseThrow(() -> new MailboxRoleNotFoundException(Role.OUTBOX));
+        } catch (MailboxException | MailboxRoleNotFoundException e) {
+            LOGGER.error("Unable to find a mailbox with role 'outbox'!");
+            throw Throwables.propagate(e);
+        }
+        return request.getCreate().entrySet().stream()
+                .map(e -> new MessageWithId.CreationMessageEntry(e.getKey(), e.getValue()))
+                .map(nuMsg -> createMessageInOutbox(nuMsg, mailboxSession, outbox, buildMessageIdFunc(mailboxSession, outbox)))
+                .reduce(SetMessagesResponse.builder(),
+                        (builder, msg) -> builder.created(ImmutableMap.of(msg.creationId, msg.message)),
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: You may define the accumulator & the combiner in SetMessagesResponse, as they are dupplicate in each processor.

File: server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java (hunk 2 above)
Comment (fvn-linagora) [inline server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java]: - omitted on purpose, as it required more refactoring work. Has been added separately in following commits, .
- default scope should be fine, as they are injected into consumer classes. No need for singletons I guess.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/MailboxRoleNotFoundException.java (hunk 4 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/exceptions/MailboxRoleNotFoundException.java]: fixed

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MessageWithId.java (hunk 1 above)
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MessageWithId.java]: made public

File: server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java (hunk 3 above)
Comment (fvn-linagora) [inline server/container/cassandra-guice/src/main/java/org/apache/james/jmap/MethodsModule.java]: Fixed. removed the provider

**Comment by fvn-linagora** [comment]
test this please

**Pull request closed without merging**